include/aos/iam/certmodule.hpp

```diff
@@ -0,0 +1,380 @@
+
+    static const Array<const char* const> GetStrings()
+    {
+        static const char* const sContentTypeStrings[] = {"clientAuth", "serverAuth"};
+        return Array<const char* const>(sContentTypeStrings, ArraySize(sContentTypeStrings));
+    };
+};
+
+using ExtendedKeyUsageEnum = ExtendedKeyUsageType::Enum;
+using ExtendedKeyUsage = EnumStringer<ExtendedKeyUsageType>;
+
+/**
+ * ModuleConfig module configuration.
+ */
+struct ModuleConfig {
+    /**
+     * Key generating algorithm.
+     */
+    KeyGenAlgorithm mKeyGenAlgorithm;
+    /**
+     * Maximum number of certificates for module.
+     */
+    unsigned mMaxCertificates = 0U;
+    /**
+     * Extra extensions needed for CSR. Current supported values: [clientAuth, serverAuth]
+     */
+    StaticArray<ExtendedKeyUsage, cModuleKeyUsagesMaxCount> mExtendedKeyUsage;
+    /**
+     * Alternative DNS names.
+     */
+    StaticArray<StaticString<crypto::cDnsNameLen>, crypto::cAltDnsNamesCount> mAlternativeNames;
+    /**
+     * Skip certificate chain validation.
+     */
+    bool mSkipValidation = false;
+};
+
+/**
+ * General certificate information.
+ */
+struct CertInfo {
+    /**
+     * Certificate issuer name.
+     */
+    StaticString<crypto::cCertificateIssuerLen> mIssuer;
+    /**
+     * Certificate serial number.
+     */
+    StaticString<crypto::cCertificateSerialNumberLen> mSerial;
+    /**
+     * Certificate url.
+     */
+    StaticString<cURLLen> mCertURL;
+    /**
+     * Certificate's private key url.
+     */
+    StaticString<cURLLen> mKeyURL;
+    /**
+     * Certificate expiration time.
+     */
+    time::Time mNotAfter;
+    /**
+     * Compares certificate info.
+     *
+     * @param certInfo info to compare.
+     * @return bool.
+     */
+    bool operator==(const CertInfo& certInfo) const;
+    /**
+     * Compares certificate info.
+     *
+     * @param certInfo info to compare.
+     * @return bool.
+     */
+    bool operator!=(const CertInfo& certInfo) const { return !operator==(certInfo); }
+};
+
+/**
+ * Platform dependent secure certificate storage.
+ */
+class SecureStorageItf {
+public:
+    /**
+     * Owns the module.
+     *
+     * @param password certficate password.
+     * @return Error.
+     */
+    virtual Error SetOwner(const String& password) = 0;
+
+    /**
+     * Removes all module certificates.
+     *
+     * @return Error.
+     */
+    virtual Error Clear() = 0;
+
+    /**
+     * Generates private key.
+     *
+     * @param password owner password.
+     * @return RetWithError<crypto::PrivateKey*>.
+     */
+    virtual RetWithError<crypto::PrivateKey*> CreateKey(const String& password, KeyGenAlgorithm algorithm) = 0;
+
+    /**
+     * Applies certificate chain to a module.
+     *
+     * @param certChain certificate chain.
+     * @param[out] topCertInfo info about a top certificate in a chain.
+     * @param[out] password owner password.
+     * @return Error.
+     */
+    virtual Error ApplyCertChain(
```

Comment:
Rename to `ApplyCert`. Also I guess we can manage keeping password in CertModule instance. No need to return it here. In this case this method should look like:
```c++
virtual RetWithError<CertInfo> ApplyCertChain(const Array<crypto::x509::Certificate>& certChain)
```